File: numass-control/vac/src/main/java/inr/numass/control/readvac/VacDeviceFactory.java
package inr.numass.control.readvac;

import hep.dataforge.context.Context;
import hep.dataforge.control.devices.Device;
import hep.dataforge.control.measurements.Sensor;
import hep.dataforge.meta.Meta;
import inr.numass.control.DeviceViewConnection;
import inr.numass.control.DeviceViewFactory;
import inr.numass.control.readvac.fx.VacCollectorView;

import java.util.List;
import java.util.stream.Collectors;

/**
 * A factory for vacuum measurements collector
 * Created by darksnake on 16-May-17.
 */
public class VacDeviceFactory implements DeviceViewFactory {
    @Override
    public String getType() {
        return "numass:vac";
    }

    public Sensor<Double> buildSensor(Context context, Meta sensorConfig) {
        switch (sensorConfig.getString("sensorType", "")) {
            case "mks":
                return new MKSVacDevice(context, sensorConfig);
            case "CM32":
                return new CM32Device(context, sensorConfig);
            case "meradat":
                return new MeradatVacDevice(context, sensorConfig);
            case "baratron":
                return new MKSBaratronDevice(context, sensorConfig);
            default:
                throw new RuntimeException("Unknown vacuum sensor type");
        }
    }

    @Override
    public VacCollectorDevice build(Context context, Meta config) {
        List<Sensor<Double>> sensors = config.getMetaList("sensor").stream()
                .map(sensorConfig -> buildSensor(context, sensorConfig)).collect(Collectors.toList());

        VacCollectorDevice collector = new VacCollectorDevice(context, config);
        collector.setSensors(sensors);
        return collector;
    }

    @Override
    public DeviceViewConnection buildView(Device device) {
        return VacCollectorView.build(device.getContext());
    }
}

File: numass-control/vac/src/main/kotlin/inr/numass/control/readvac/VacDeviceFactory.kt
package inr.numass.control.readvac

import hep.dataforge.context.Context
import hep.dataforge.control.devices.DeviceFactory
import hep.dataforge.control.devices.Sensor
import hep.dataforge.meta.Meta
import java.util.stream.Collectors

/**
 * A factory for vacuum measurements collector
 * Created by darksnake on 16-May-17.
 */
class VacDeviceFactory : DeviceFactory {
    override val type: String = "numass.vac"

    private fun buildSensor(context: Context, sensorConfig: Meta): Sensor {
        return when (sensorConfig.getString("sensorType", "")) {
            "mks" -> MKSVacDevice(context, sensorConfig)
            "CM32" -> CM32Device(context, sensorConfig)
            "meradat" -> MeradatVacDevice(context, sensorConfig)
            "baratron" -> MKSBaratronDevice(context, sensorConfig)
            "ThyroCont" -> ThyroContVacDevice(context,sensorConfig)
//            VIRTUAL_SENSOR_TYPE -> VirtualDevice.randomDoubleSensor(context, sensorConfig)
            else -> throw RuntimeException("Unknown vacuum sensor type")
        }
    }

    override fun build(context: Context, config: Meta): VacCollectorDevice {
        val sensors = config.getMetaList("sensor").stream()
                .map { sensorConfig -> buildSensor(context, sensorConfig) }
                .collect(Collectors.toList<Sensor>())

        return VacCollectorDevice(context, config, sensors)
    }

//    override fun buildView(device: Device): DeviceDisplayFX<VacCollectorDevice> {
//        return VacCollectorDisplay();
//    }
}
